src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading.Tasks;
+
+namespace System.Threading.RateLimiting
+{
+    /// <summary>
+    /// Contains methods to assist with creating a <see cref="PartitionedRateLimiter{TResource}"/>.
+    /// </summary>
+    public static class PartitionedRateLimiter
+    {
+        /// <summary>
+        /// Method used to create a default implementation of <see cref="PartitionedRateLimiter{TResource}"/>.
+        /// </summary>
+        /// <typeparam name="TResource">The resource type that is being rate limited.</typeparam>
+        /// <typeparam name="TPartitionKey">The type to distinguish partitions with.</typeparam>
+        /// <param name="partitioner">Method called every time an Acquire or WaitAsync call is made to figure out what rate limiter to apply to the request.
+        /// If the <see cref="RateLimitPartition{TKey}.PartitionKey"/> matches a cached entry then the rate limiter previously used for that key is used. Otherwise, the factory is called to get a new rate limiter.</param>
+        /// <param name="equalityComparer">Optional <see cref="IEqualityComparer{T}"/> to customize the comparison logic for <typeparamref name="TPartitionKey"/>.</param>
+        /// <returns></returns>
+        public static PartitionedRateLimiter<TResource> Create<TResource, TPartitionKey>(
+            Func<TResource, RateLimitPartition<TPartitionKey>> partitioner,
+            IEqualityComparer<TPartitionKey>? equalityComparer = null) where TPartitionKey : notnull
+        {
+            return new DefaultPartitionedRateLimiter<TResource, TPartitionKey>(partitioner, equalityComparer);
+        }
+    }
+
+    internal sealed class DefaultPartitionedRateLimiter<TResource, TKey> : PartitionedRateLimiter<TResource> where TKey : notnull
+    {
+        private readonly Func<TResource, RateLimitPartition<TKey>> _partitioner;
+
+        // TODO: Look at ConcurrentDictionary to try and avoid a global lock
+        private Dictionary<TKey, Lazy<RateLimiter>> _limiters;
+        private Timer? _timer;
+        private bool _disposed;
+
+        // Use the Dictionary as the lock field so we don't need to allocate another object for a lock and have another field in the object
+        private object Lock => _limiters;
+
+        public DefaultPartitionedRateLimiter(Func<TResource, RateLimitPartition<TKey>> partitioner,
+            IEqualityComparer<TKey>? equalityComparer = null)
+        {
+            _limiters = new Dictionary<TKey, Lazy<RateLimiter>>(equalityComparer);
+            _partitioner = partitioner;
+
+            // TODO: Only create timer once there is an active replenishing limiter
+            // TODO: Figure out what interval we should use
```

Comment:
Can we file issues for the any TODOs we think are worth following up in the .NET 7 timeframe? I think these probably qualify. I think we might want to add public API to make this interval configurable.